common/transport/amqp/src/amqp.ts

```diff
@@ -135,6 +88,280 @@ export class Amqp {
+                });
+                debug('sender link attached: ' + endpoint);
+                done(null, this._receivers[endpoint]);
+              }
+            });
+          },
+          attachSenderLink: (endpoint: string, linkOptions: any, done: GenericAmqpBaseCallback<any>): void => {
+            debug('creating sender link for: ' + endpoint);
+            let senderFsm = new SenderLink(endpoint, linkOptions, this._amqp);
+            this._senders[endpoint] = senderFsm;
+            const attachErrorHandler = (err) => {
+              debug('sender link error event while attaching: ' + endpoint + ': ' + err.toString());
+              delete(this._senders[endpoint]);
+              done(err);
+            };
+
+            this._senders[endpoint].on('error', attachErrorHandler);
+            debug('attaching sender link for: ' + endpoint);
+            this._senders[endpoint].attach((err) => {
+              if (err) {
+                debug('failed to attach sender link: ' + endpoint + ': ' + err.toString());
+                done(err);
+              } else {
+                this._senders[endpoint].removeListener('error', attachErrorHandler);
+                this._senders[endpoint].on('error', (err) => {
+                  debug('error on sender link: ' + endpoint + ': ' + err.toString());
+                  delete(this._senders[endpoint]);
+                });
+                debug('sender link attached: ' + endpoint);
+                done(null, this._senders[endpoint]);
+              }
+            });
+          },
+          detachReceiverLink: (endpoint: string, detachCallback: GenericAmqpBaseCallback<any>): void => {
+            if (!this._receivers[endpoint]) {
+              this._safeCallback(detachCallback);
+            } else {
+              this._detachLink(this._receivers[endpoint], (err?) => {
+                delete(this._receivers[endpoint]);
+                this._safeCallback(detachCallback, err);
+              });
+            }
+          },
+          detachSenderLink: (endpoint: string, detachCallback: GenericAmqpBaseCallback<any>): void => {
+            this._detachLink(this._senders[endpoint], (err?) => {
+              delete(this._senders[endpoint]);
+              this._safeCallback(detachCallback, err);
+            });
+          }
+        },
+        disconnecting: {
+          _onEnter: (disconnectCallback, err) => {
+            const disconnectFunc = err ? (callback) => {
+              callback();
+            } : (callback) => {
+              this._amqp.removeAllListeners('disconnected');
+              /*Codes_SRS_NODE_COMMON_AMQP_16_004: [The disconnect method shall call the done callback when the application/service has been successfully disconnected from the service] */
+              this._amqp.disconnect().then(() => {
+                callback();
+                return null;
+              }).catch((err) => {
+                callback(err);
+              });
```

Comment:
If we're disconnecting because of an error, don't we want to do this?  Otherwise, the connection seems like it might be only partially taken down.

Reply:
would be fine if we could rely on amqp10 to do the right thing if the connection is already disconnected or partially disconnected, but it crashes. so we have to assume that if there is an error, amqp10 already took down the connection and we can't call disconnect again.

---
In reply to: [139531957](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139531957) [](ancestors = 139531957)